tally/src/noop_stats_reporter.cc

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "tally/src/noop_stats_reporter.h"
+
+#include "tally/src/capable_of.h"
+
+namespace tally {
+
+std::shared_ptr<StatsReporter> NoopStatsReporter::New() {
```

Comment:
So I have done the `Noop` thing too in Jaeger and then thought people may not like it. In this case it might be worse because you aren't inlining the noops in the class declaration header. The compiler may not realize these are noops and incur the function call overhead unnecessarily. Definitely suggest moving any empty function definition to header. There may be a way around `Noop` objects entirely involving macros, but I haven't thought of a good solution yet.

Reply:
Fortunately this is really just used internally and isn't exposed publicly so I think it's okay for now.